### PR TITLE
Move IO types into System.Runtime

### PR DESCRIPTION
**Pull request by JeremyKuhne**
Move all of System.IO and System.IO.Primitives
into System.Runtime.

@weshaggard, @ianhays

### CODE REVIEW

**Comment by jkotas** [comment]
I think the TextReader/Writer, BinaryReader/Writer and friends should be rather moved to System.Runtime.Extensions. In particular, System\CodeDom\Compiler\IndentedTextWriter.cs is pretty bad candidate for System.Runtime.

**Comment by JeremyKuhne** [comment]
@jkotas, @weshaggard My understanding is that S.R.E is not intended to live on.

**Comment by weshaggard** [comment]
System.Runtime.Extensions will live on in some form or fashion. We don't plan to completely collapse it into System.Runtime.

**Comment by weshaggard** [comment]
I think I would move all of these into System.Runtime.Extensions as have them be implemented there.

**Comment by JeremyKuhne** [comment]
@weshaggard; @jkotas  What is the split? Can we come up with some sort of guidance on this? I understand that we need to keep SRE for compat, but what is the functional difference?

**Comment by jkotas** [comment]
My model is that System.Runtime.Extensions is for compat types (if possible due to dependencies): AppDomain, remoting, low-perf constructs, ... .

TextReader/Writer, BinaryReader/Writer are examples of the low-perf constructs - all methods are virtual, allocate a lot more than you would like, ... .

**Comment by weshaggard** [comment]
While this isn't strictly true my general model is things in System.Runtime are generally implemented in CoreLib and things in System.Runtime.Extensions are generally implemented directly inside System.Runtime.Extensions. Given most of System.IO is in the case of shared implementation I would keep them in System.Runtime.Extensions.

**Comment by JeremyKuhne** [comment]
I'm finding it hard to see what the advantage is to consuming code. Is there ever really a case where you can get away with just a reference to S.R. and not S.R.E? I'm a little worried that we're not crisp here- that we might be trying to come up with a distinction just to satisfy the fact that we already had a S.R.E. and that we could end up (long term) with a bunch of type forwards from S.R.E. to S.R.

That said, I'll update the PR as soon as I can.

**Comment by weshaggard** [comment]
>I'm finding it hard to see what the advantage is to consuming code. Is there ever really a case where you can get away with just a reference to S.R. and not S.R.E? 

For real apps people will likely need to reference a lot more then just these but that doesn't mean we should put everything into System.Runtime. We still want some level of factoring that doesn't entail everything going into the core assembly. I think both the reasons @jkotas and I stated are good reasons to keep things isolated into another assembly.

**Comment by JeremyKuhne** [comment]
> that doesn't mean we should put everything into System.Runtime

I'm not trying to suggest that. I'm just concerned. :) If we got serious about S.R. being dependent on CoreLib, and make S.R.E CoreLib free I'd feel better I think. Then I can see some real value- e.g. S.R.E is reusable.

**Comment by JeremyKuhne** [comment]
Note that I'm blocking others on this change. @joperezr

File: src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj [hunk 1]
```diff
@@ -91,6 +110,17 @@
       <Link>Common\System\HResults.cs</Link>
     </Compile>
   </ItemGroup>
+  <ItemGroup Condition="'$(TargetGroup)' == ''">
+    <Compile Include="System\IO\FileStreamHelpers.CoreClr.cs" />
+    <Compile Include="System\IO\TextReader.netstandard17.cs" />
+    <Compile Include="System\IO\TextWriter.netstandard17.cs" />
+    <ProjectReference Include="..\..\System.Diagnostics.Debug\src\System.Diagnostics.Debug.csproj">
+      <Project>{E7E8DE8A-9EC1-46A8-A6EE-727DB32DBEB8}</Project>
```
Comment (weshaggard) [inline src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj]: Can you please group the ProjectReference with the other references in this project? I would like to keep them mostly clustered together.

Comment (weshaggard) [inline src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj]: In fact it looks like this ProjectReference already exists.

**Comment by weshaggard** [comment]
Looks like the conflicts are causing dependency validation failures.

File: src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj (hunk 1 above)
Comment (JeremyKuhne) [inline src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj]: Fixed- picked it up integrating.

File: src/System.IO/src/System.IO.csproj [hunk 2]
```diff
@@ -24,71 +24,16 @@
...
-    </Compile>
-    <Compile Include="$(Commonpath)\System\SR.Core.cs">
-      <Link>Common\System\SR.Core.cs</Link>
-    </Compile>
-  </ItemGroup>
   <ItemGroup Condition="'$(TargetGroup)' != 'net462' and '$(TargetGroup)' != 'net463'">
-    <TargetingPackReference Include="System.Private.CoreLib" />
-    <TargetingPackReference Include="System.Private.Interop" Condition="'$(TargetGroup)' == 'uap101aot'" />
+    <!-- ToDo: Remove this P2P reference once new packages are produced and updated -->
+    <ProjectReference Include="..\..\System.Runtime\pkg\System.Runtime.pkgproj" />
```
Comment (weshaggard) [inline src/System.IO/src/System.IO.csproj]: I'm surprised that this actually works. In general pkgproj references only work in test projects. I'm find with this temporarily if it works though.

Comment (JeremyKuhne) [inline src/System.IO/src/System.IO.csproj]: Heh, well I guess my ignorance is bliss in this case. :)

File: src/System.Runtime.WindowsRuntime/src/System.Runtime.WindowsRuntime.csproj [hunk 3]
```diff
@@ -65,9 +66,6 @@
     <ProjectReference Include="..\..\System.IO\src\System.IO.csproj">
       <TargetGroup>uap101aot</TargetGroup>
     </ProjectReference>
-    <ProjectReference Include="..\..\System.Runtime\src\System.Runtime.csproj">
```
Comment (weshaggard) [inline src/System.Runtime.WindowsRuntime/src/System.Runtime.WindowsRuntime.csproj]: Why was this removed? I would expected it to still be necessary. In fact I would kind of expect the System.IO.csproj reference to need to be updated to System.Runtime.Extensions.

Comment (JeremyKuhne) [inline src/System.Runtime.WindowsRuntime/src/System.Runtime.WindowsRuntime.csproj]: I was hitting multiple type conflicts, removing this one fixed a number of them
